capsules/core/src/alarm.rs

```diff
@@ -17,140 +17,339 @@ use crate::driver;
-
-                    earliest_alarm = match earliest_alarm {
-                        Expiration::Disabled => {
-                            earliest_end = current_end_ticks;
-                            alarm.expiration
-                        }
-                        Expiration::Enabled { reference, dt } => {
-                            // There are two cases when current might be
-                            // an earlier alarm.  The first is if it
-                            // fires inside the interval (reference,
-                            // reference+dt) of the existing earliest.
-                            // The second is if now is not within the
-                            // interval: this means that it has
-                            // passed. It could be the earliest has passed
-                            // too, but at this point we don't need to track
-                            // which is earlier: the key point is that
-                            // the alarm must fire immediately, and then when
-                            // we handle the alarm callback the userspace
-                            // callbacks will all be pushed onto processes.
-                            // Because there is at most a single callback per
-                            // process and they must go through the scheduler
-                            // we don't care about the order in which we push
-                            // their callbacks, as their order of execution is
-                            // determined by the scheduler not push order. -pal
-                            let temp_earliest_reference = A::Ticks::from(reference);
-                            let temp_earliest_dt = A::Ticks::from(dt);
-                            let temp_earliest_end =
-                                temp_earliest_reference.wrapping_add(temp_earliest_dt);
-
-                            if current_end_ticks
-                                .within_range(temp_earliest_reference, temp_earliest_end)
-                            {
-                                earliest_end = current_end_ticks;
-                                alarm.expiration
-                            } else if !now_lower_bits
-                                .within_range(temp_earliest_reference, temp_earliest_end)
-                            {
-                                earliest_end = temp_earliest_end;
-                                alarm.expiration
-                            } else {
-                                earliest_alarm
-                            }
-                        }
+    // TODO: document
+    fn earliest_alarm<UD, R, F: FnOnce(Expiration<A::Ticks>, &UD) -> Option<R>>(
+        now: A::Ticks,
+        expirations: impl Iterator<Item = (Expiration<A::Ticks>, UD, F)>,
+        // expired_handler: &mut impl FnMut(&Expiration<A::Ticks>, &UD) -> Option<R>,
+    ) -> Result<Option<(Expiration<A::Ticks>, UD)>, (Expiration<A::Ticks>, UD, R)> {
```

Comment:
This function does much of the heavy lifting of re-arming the timer, as it's generic. This also makes it hard to understand. It is in desperate need of documentation. `expired_handler` as a parameter has been integrated into `expirations` and should be removed.

Reply:
I believe this misses a corner case where:

```
now = 1000
expirations[0].fire_at = 1050
expirations[1].fire_at = 950
```

if the `expirations[1].expired_handler` runs for >50 ticks, then by the time `expirations[0]` is processed as the "earliest alarm" in the block at line 193, its `dt` will look far in the future.

Reply:
I might well be wrong, but I don't think that this is a concern here.

First of all, this function takes `now` as a parameter. We do this primarily to avoid performing a volatile read of the underlying timer's register multiple times, but it also ensures that this function and all of the expired-callbacks are executed from a consistent reference time. This means that, regardless of how long each expired handler takes, we always identify the earliest alarm from this reference time:

https://github.com/tock/tock/blob/c8e4da7bc1f7648c08eda48b691e1c76eb554254/capsules/core/src/alarm.rs#L139-L190

Apart from that, we also don't just save the `fire_at` value, but `reference` and `dt`, where `fire_at = reference.wrapping_add(dt)`. This provides us with significant margin of error, provided that `dt` is significantly smaller than `Ticks::max()`. Even if we were to take long, once `now` surpasses `reference + dt` and did not tick another `Ticks::max() - dt`, we'll still treat it as expired. We also don't risk mistaking a previous timer as the "earliest", as when `now` is not within `[reference; dt)`, we treat is as expired instead.

This does raise another question though: I believe that all of this falls apart when `Ticks::max() - dt` is very small. In this case, we'd run a realistic risk of _never_ firing a timer, as we'd never trigger the case where `now` is within `[reference; dt)`.

Should we limit the maximum value of `dt`, perhaps to `Ticks::max() / 2`, to ensure that we always reliably detect when it has expired and have enough time to schedule the callback?

Reply:
Ah, okay, I think lost in my reading through the web UI diff view was where the `reference, dt` was coming from that is ultimately passed to `set_alarm`.

The concern does come back up in `set_alarm` implementations, as they all _do_ do a fresh read of `now`. I looked through several (chosen arbitrarily as example, the [rp2040 alarm implementation](https://github.com/tock/tock/blob/e7ab99bf64b6055d40a9de2bbd6fd0cc459ed622/chips/rp2040/src/timer.rs#L241) does this right) and all seem to handle the case where `now` isn't in the [ref,dt] range. _Most_ seem to also account for `minimum_dt()` in the `set_alarm` implementation, but it looks like [esp32](https://github.com/tock/tock/blob/master/chips/esp32/src/timg.rs#L237) might be missing that nuance. Not a bug for this PR though.